prometheus/promhttp/instrument_server.go

```diff
@@ -0,0 +1,433 @@
+
+	if code {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			d := newDelegator(w)
+			next.ServeHTTP(d, r)
+			size := computeApproximateRequestSize(r)
+			obs.With(labels(code, method, r.Method, d.Status())).Observe(float64(size))
+		})
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		next.ServeHTTP(w, r)
+		size := computeApproximateRequestSize(r)
+		obs.With(labels(code, method, r.Method, 0)).Observe(float64(size))
+	})
+}
+
+// InstrumentHandlerResponseSize accepts an ObserverVec interface and an
+// http.Handler, returning a new http.Handler that wraps the supplied
+// http.Handler. The provided ObserverVec must be registered in a registry in
+// order to be used. If the wrapped http.Handler has not set a status code,
+// i.e. the value is currently 0, the supplied ObserverVec will report a 200.
+// The instance labels "code" and "method" are supported on the provided
+// ObserverVec.
+// Note: Partitioning histograms is expensive.
+func InstrumentHandlerResponseSize(obs prometheus.ObserverVec, next http.Handler) http.Handler {
+	code, method := checkLabels(obs)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		d := newDelegator(w)
+		next.ServeHTTP(d, r)
+		obs.With(labels(code, method, r.Method, d.Status())).Observe(float64(d.Written()))
+	})
+}
+
+func checkLabels(c prometheus.Collector) (code bool, method bool) {
+	// TODO(beorn7): Remove this hacky way to check for instance labels
+	// once Descriptors can have their dimensionality queried.
+	var (
+		desc *prometheus.Desc
+		pm   dto.Metric
+	)
+
+	descc := make(chan *prometheus.Desc, 1)
+	c.Describe(descc)
+
+	select {
+	case desc = <-descc:
+	default:
+		panic("no description provided by collector")
+	}
+	select {
+	case <-descc:
+		panic("more than one description provided by collector")
+	default:
+	}
+
+	close(descc)
+
+	if _, err := prometheus.NewConstMetric(desc, prometheus.UntypedValue, 0); err == nil {
+		return
+	} else if m, err := prometheus.NewConstMetric(desc, prometheus.UntypedValue, 0, ""); err == nil {
+		if err := m.Write(&pm); err != nil {
+			panic("error checking metric for labels")
+		}
+
+		name := *pm.Label[0].Name
+		if name == "code" {
+			code = true
+		} else if name == "method" {
+			method = true
+		} else {
+			panic("metric partitioned with non-supported labels")
+		}
+		return
+	} else if m, err := prometheus.NewConstMetric(desc, prometheus.UntypedValue, 0, "", ""); err == nil {
+		if err := m.Write(&pm); err != nil {
+			panic("error checking metric for labels")
+		}
+
+		for _, label := range pm.Label {
+			if *label.Name == "code" || *label.Name == "method" {
+				continue
+			}
+			panic("metric partitioned with non-supported labels")
+		}
+
+		code = true
+		method = true
+		return
+	}
+
+	panic("metric partitioned with non-supported labels")
+}
+
+func labels(code, method bool, reqMethod string, status int) prometheus.Labels {
+	labels := prometheus.Labels{}
```

Comment:
This creates a new map on each call of `labels` (even if no partitioning is requested at all).

Suggestion: Create a map in the middleware and pass it in into the labels function. Since the label names are always the same, you don't have to ever clean the map, just update code and/or method here.